provider/telegram.go

```diff
@@ -0,0 +1,395 @@
+
+// LogoutHandler - GET /logout
+func (t *TelegramHandler) LogoutHandler(w http.ResponseWriter, r *http.Request) {
+	t.TokenService.Reset(w)
+}
+
+// tgAPI implements TelegramAPI
+type tgAPI struct {
+	logger.L
+	endpoint     string
+	token        string
+	updateOffset int
+}
+
+// NewTelegramAPI returns initialized TelegramAPI implementation
+func NewTelegramAPI(token string, l logger.L) TelegramAPI {
+	return &tgAPI{
+		L:        l,
+		endpoint: "https://api.telegram.org",
+		token:    token,
+	}
+}
+
+// GetUpdates fetches incoming updates
+func (t *tgAPI) GetUpdates(ctx context.Context) (*telegramUpdate, error) {
+	url := `getUpdates?allowed_updates=["message"]`
+	if t.updateOffset != 0 {
+		url += fmt.Sprintf("&offset=%d", t.updateOffset) // See core.telegram.org/bots/api#getupdates
+	}
+
+	var result telegramUpdate
+
+	err := t.request(ctx, url, &result)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to fetch updates")
+	}
+
+	for _, u := range result.Result {
+		if u.UpdateID >= t.updateOffset {
+			t.updateOffset = u.UpdateID + 1
+		}
+	}
+
+	return &result, err
+}
+
+// Send sends a message to telegram peer
+func (t *tgAPI) Send(ctx context.Context, id int, msg string) error {
+	url := fmt.Sprintf("%s/bot%s/sendMessage?chat_id=%d&text=%s", t.endpoint, t.token, id, neturl.PathEscape(msg))
+
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
+	if err != nil {
+		return errors.Wrap(err, "failed to create request")
+	}
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed to send request")
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return errors.Errorf("telegram returned %d status code", resp.StatusCode)
+	}
+
+	return nil
+}
+
+// Avatar returns URL to user avatar
+func (t *tgAPI) Avatar(ctx context.Context, id int) (string, error) {
+	// Get profile pictures
+	url := fmt.Sprintf(`getUserProfilePhotos?user_id=%d`, id)
+
+	var profilePhotos = struct {
+		Result struct {
+			Photos [][]struct {
+				ID string `json:"file_id"`
+			} `json:"photos"`
+		} `json:"result"`
+	}{}
+
+	err := t.request(ctx, url, &profilePhotos)
+	if err != nil {
+		return "", err
+	}
+
+	// User does not have profile picture set or it is hidden in privacy settings
+	if len(profilePhotos.Result.Photos) == 0 {
+		return "", nil
+	}
+
+	// Get actual avatar url
+	last := len(profilePhotos.Result.Photos[0]) - 1
+	fileID := profilePhotos.Result.Photos[0][last].ID
+	url = fmt.Sprintf(`getFile?file_id=%s`, fileID)
+
+	var fileMetadata = struct {
+		Result struct {
+			Path string `json:"file_path"`
+		} `json:"result"`
+	}{}
+
+	err = t.request(ctx, url, &fileMetadata)
+	if err != nil {
+		return "", err
+	}
+
+	avatarURL := fmt.Sprintf("%s/file/bot%s/%s", t.endpoint, t.token, fileMetadata.Result.Path)
+
+	return avatarURL, nil
+}
+
+func (t *tgAPI) request(ctx context.Context, method string, data interface{}) error {
```

Comment:
I'm not familiar enough with telegram API, but I would guess adding some retry if the call failed makes some sense here. If you agree take a look on github.com/go-pkgz/repeater